### PR TITLE
HomeApp: webSocket connection for notification of active card

### PR DESCRIPTION
**Pull request by TomaszLukawskiSam**
[Issue] https://github.com/Samsung/TAU/issues/1417
[Problem] Implement notification mechanism
[Solution]
  - websocket connection for notification from HomeTVAppServer
  - added synchronization of local apps and remote apps
  - change apps order

Signed-off-by: Tomasz Lukawski <t.lukawski@samsung.com>

### CODE REVIEW

File: examples/mobile/HomeApp/js/app.js [hunk 1]
```diff
@@ -11,19 +11,36 @@ import Storage from "./clipping-storage.js";
 			"webclip/tv-remote-control",
 			"webclip/restaurant",
 			"webclip/weather"
-		];
+		],
+		wsPort = 8080,
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js]: on 8080 we run httpserver, how about using other port?

**Comment by lslachciak** [comment]
I'm observing ~~2~~ 1 issue~~s~~ in in the following scenario:

1. Clean local storage
2. Run HomeAppServer
3. Open HomeApp
~~**[Observe]: Nothing is visible, but some webclips should be selected by default**~~
(deleted by default server is not sending isSelected)
4. Click Add button
5. Select some webclips. 
6. Select OK
7. Refresh browser
**[Observe] Errors on console:**
```
ncaught TypeError: Cannot read property 'replaceChild' of null
    at Card.prototype.changeContent (tau.js:28779)
    at CardRoute.prototype.open (tau.js:60839)
    at Object.resolve (tau.js:15709)
    at Router._loadSuccess (tau.js:16383)
    at tau.js:16329
    at targetCallback (tau.js:15245)
    at templateCallback (tau.js:15250)
    at callbackFunction (tau.js:60891)

```

File: examples/mobile/HomeApp/js/app.js [hunk 2]
```diff
@@ -224,17 +309,18 @@ import Storage from "./clipping-storage.js";
...
-		if (!appsList.length) {
-			getAppsList.then((apps) => {
-				appsList = apps;
-				updateWebClipsUI();
-				updateWebClipListPopup();
-				storage.refreshStorage(Storage.elements.APPSLIST, apps);
-			});
-		} else {
+		if (appsList.length) {
 			updateWebClipsUI();
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js]: those two functions will be called again later,
do we really want to call it twice?

File: examples/mobile/HomeAppServer/app.js [hunk 3]
```diff
@@ -1,64 +1,160 @@
...
+			{
+				url: "webclip/weather"
+			}
+		]
+	}];
 
 app.use(logger("dev"));
 app.use(express.json());
 app.use(express.urlencoded({ extended: false }));
 app.use(cookieParser());
```
Comment (lslachciak) [inline examples/mobile/HomeAppServer/app.js:55]: do we use cookies?

Comment (TomaszLukawskiSam) [reply examples/mobile/HomeAppServer/app.js:55]: I don't know. This is your code.

Comment (TomaszLukawskiSam) [reply examples/mobile/HomeAppServer/app.js:55]: We can left cookie. Cookies are useful for user session

File: examples/mobile/HomeApp/js/app.js (hunk 2 above)
Comment (TomaszLukawskiSam) [reply examples/mobile/HomeApp/js/app.js]: This code is twice but first code is synchronous and second code is asynchronous and may be called much later.

**Comment by TomaszLukawskiSam** [comment]
> I'm observing ~2~ 1 issue~s~ in in the following scenario:
> 
>     1. Clean local storage
> 
>     2. Run HomeAppServer
> 
>     3. Open HomeApp
>        ~**[Observe]: Nothing is visible, but some webclips should be selected by default**~
>        (deleted by default server is not sending isSelected)
> 
>     4. Click Add button
> 
>     5. Select some webclips.
> 
>     6. Select OK
> 
>     7. Refresh browser
>        **[Observe] Errors on console:**
> 
> 
> ```
> ncaught TypeError: Cannot read property 'replaceChild' of null
>     at Card.prototype.changeContent (tau.js:28779)
>     at CardRoute.prototype.open (tau.js:60839)
>     at Object.resolve (tau.js:15709)
>     at Router._loadSuccess (tau.js:16383)
>     at tau.js:16329
>     at targetCallback (tau.js:15245)
>     at templateCallback (tau.js:15250)
>     at callbackFunction (tau.js:60891)
> ```

I fixed this issue.

**Comment by lslachciak** [comment]
In this PR when having connnection to HomeAppServer I'm geeting errors like
`Cannot read property 'querySelectorAl`
when no connection to HomeAppsErver additional errors like:
` Cannot read property 'replaceChild' appear`

**Comment by TomaszLukawskiSam** [comment]
> In this PR when having connnection to HomeAppServer I'm geeting errors like
> `Cannot read property 'querySelectorAl`
> when no connection to HomeAppsErver additional errors like:
> ` Cannot read property 'replaceChild' appear`

Did you rebuild TAU?

File: examples/mobile/HomeApp/js/app.js [hunk 4]
```diff
@@ -6,80 +6,168 @@ import Storage from "./clipping-storage.js";
...
+					resolve(apps);
 				})
 		});
 
+	function updateAppsList(apps) {
+		var change = false,
+			appsCount = appsList.length,
+			currentOrder = "";
+
+		// remove remote app from local app list
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js]: comment description is misleading I think.
I understand that you leave only (not remove) apps which have the same id as those from remote

File: examples/mobile/HomeApp/js/app.js [hunk 5]
```diff
@@ -6,80 +6,168 @@ import Storage from "./clipping-storage.js";
...
+					}
+					resolve(apps);
 				})
 		});
 
+	function updateAppsList(apps) {
+		var change = false,
+			appsCount = appsList.length,
+			currentOrder = "";
+
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js:75]: wouldn't it be easier to create list of selected webclips with their app ids, make an assignement
appsList = apps
and update selected property only?

File: examples/mobile/HomeApp/js/app.js (hunk 4 above)
Comment (TomaszLukawskiSam) [reply examples/mobile/HomeApp/js/app.js]: Ok, I'll change it

File: examples/mobile/HomeApp/js/app.js (hunk 5 above)
Comment (TomaszLukawskiSam) [reply examples/mobile/HomeApp/js/app.js:75]: If we plan to add information from the manifest, then will not be a good option.

**Comment by lslachciak** [comment]
> Did you rebuild TAU?

No. My fault.
Although after TAU rebuild I got warning
`Card: Card cannot be updated. Parent element of card is detached from DOM`

**Comment by TomaszLukawskiSam** [comment]
> > Did you rebuild TAU?
> 
> No. My fault.
> Although after TAU rebuild I got warning
> `Card: Card cannot be updated. Parent element of card is detached from DOM`

This is the use case when developer remove service card before this card fully loaded. I added this message but I can remove it.

**Review by lslachciak — APPROVED** [review]
Ok I approve it but in nearest future we need to think about
- allowing to configure port number for WS connection
- replacing sending/receiving whole applist in favor to change only

File: examples/mobile/HomeApp/js/app.js [hunk 6]
```diff
@@ -220,21 +312,25 @@ import Storage from "./clipping-storage.js";
...
-			});
-		} else {
+		if (appsList.length) {
 			updateWebClipsUI();
 			updateWebClipListPopup();
 		}
+		// check webclips on remote server
+		getAppsList.then((apps) => {
+			updateAppsList(apps);
+			storage.refreshStorage(Storage.elements.APPSLIST, appsList);
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js:319]: you write to storage appsList but updateAppsList with apps

File: examples/mobile/HomeApp/js/app.js [hunk 7]
```diff
@@ -6,80 +6,170 @@ import Storage from "./clipping-storage.js";
...
-					resolve(data);
+					addWSListener(data.wsPort);
+					resolve(data.apps);
 				})
 				.catch(() => {
-					// use default apps list if sth wrong
-					resolve(defaultList);
+					// use apps list from storage or default apps list if sth wrong
+					let apps = storage.readAllFromStorage(Storage.elements.APPSLIST);
+
```
Comment (lslachciak) [inline examples/mobile/HomeApp/js/app.js]: appsList already contain storage or default list in this case
how about just rejecting?

**Review by lslachciak — CHANGES_REQUESTED** [review]
I'm sorry but I think I found 1 issue and one doubt regarding your PR, so I need to change my review

File: examples/mobile/HomeApp/js/app.js (hunk 6 above)
Comment (TomaszLukawskiSam) [reply examples/mobile/HomeApp/js/app.js:319]: This is properly. The method `updateAppsList` changes `appsList` variable so after update I save new state of `appsList` to storage

Comment (lslachciak) [reply examples/mobile/HomeApp/js/app.js:319]: ok, this was a little bit misleading for me, but I see your point

**Review by lslachciak — APPROVED** [review]
After clarification comment regarding 1 issue now I can approve this PR
My second comment can be treated as a future improvement

**Pull request merged**